File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/battle/BattleContext.java
package net.bestia.zoneserver.battle;

import java.util.Objects;

import net.bestia.entity.Entity;
import net.bestia.model.domain.Attack;
import net.bestia.model.domain.ConditionValues;
import net.bestia.model.domain.Element;
import net.bestia.model.domain.StatusPoints;
import net.bestia.model.entity.StatusBasedValues;

/**
 * Data transfer object to carry all needed data during a damage calculation.
 * 
 * @author Thomas Felix
 *
 */
public final class BattleContext {

	public static final class Builder {

		private final Attack usedAttack;
		private final Entity attacker;
		private final DamageVariables damageVariables;
		
		private Entity defender;
		
		private StatusPoints attackerStatusPoints;
		private StatusPoints defenderStatusPoints;
		
		private StatusBasedValues attackerStatusBased;
		private StatusBasedValues defenderStatusBased;
		
		private ConditionValues attackerCondValues;
		private ConditionValues defenderCondValues;

		public Builder(Attack usedAttack, Entity attacker, DamageVariables dmgVars) {

			this.usedAttack = Objects.requireNonNull(usedAttack);
			this.attacker = Objects.requireNonNull(attacker);
			this.damageVariables = Objects.requireNonNull(dmgVars);
		}

		public BattleContext build() {
			return new BattleContext(this);
		}

		public Builder setAttackerStatus(StatusPoints atkStatus) {
			this.attackerStatusPoints = atkStatus;
			return this;
		}

		public Builder setDefenderStatus(StatusPoints defStatus) {
			this.defenderStatusPoints = defStatus;
			return this;
		}

		public Builder setAttackerBasedValues(StatusBasedValues atkStatusBased) {
			this.attackerStatusBased = atkStatusBased;
			return this;
		}
		
		public Builder setDefenderBasedValues(StatusBasedValues defStatusBased) {
			this.defenderStatusBased = defStatusBased;
			return this;
		}

		public Builder setAttackerCondition(ConditionValues atkCond) {
			this.attackerCondValues = atkCond;
			return this;
		}

		public Builder setDefenderCondition(ConditionValues defCond) {
			this.defenderCondValues = defCond;
			return this;
		}

	}

	private final Attack usedAttack;
	private final Entity attacker;
	private final DamageVariables damageVariables;

	private final Entity defender;

	private final StatusPoints attackerStatusPoints;
	private final StatusPoints defenderStatusPoints;
	private final StatusBasedValues attackerStatusBased;
	private final StatusBasedValues defenderStatusBased;
	private final ConditionValues attackerCondtitionValues;
	private final ConditionValues defenderConditionValues;

	private BattleContext(Builder builder) {

		this.usedAttack = builder.usedAttack;
		this.attacker = builder.attacker;
		this.damageVariables = builder.damageVariables;

		this.defender = builder.defender;

		this.attackerStatusPoints = builder.attackerStatusPoints;
		this.defenderStatusPoints = builder.defenderStatusPoints;
		this.attackerStatusBased = builder.attackerStatusBased;
		this.defenderStatusBased = builder.defenderStatusBased;
		this.attackerCondtitionValues = builder.attackerCondValues;
		this.defenderConditionValues = builder.defenderCondValues;
	}

	public Attack getUsedAttack() {
		return usedAttack;
	}

	public Entity getAttacker() {
		return attacker;
	}

	public Entity getDefender() {
		return defender;
	}
	
	public DamageVariables getDamageVariables() {
		return damageVariables;
	}

	public StatusPoints getAttackerStatusPoints() {
		return attackerStatusPoints;
	}

	public StatusPoints getDefenderStatusPoints() {
		return defenderStatusPoints;
	}

	public StatusBasedValues getAttackerStatusBased() {
		return attackerStatusBased;
	}

	public StatusBasedValues getDefenderStatusBased() {
		return defenderStatusBased;
	}

	public ConditionValues getAttackerCondition() {
		return attackerCondtitionValues;
	}
	
	public ConditionValues getDefenderCondition() {
		return defenderConditionValues;
	}

	public Element getDefenderElement() {
		// TODO Auto-generated method stub
		return null;
	}

	public Element getAttackElement() {
		// TODO Auto-generated method stub
		return null;
	}

	public int getAttackerLevel() {
		// TODO Auto-generated method stub
		return 0;
	}

	public int getDefenderLevel() {
		// TODO Auto-generated method stub
		return 0;
	}
}

File: zone-server/src/main/kotlin/net/bestia/zone/battle/BattleContext.kt
package net.bestia.zone.battle

import net.bestia.zone.battle.skill.BattleAttack
import net.bestia.zone.battle.damage.DamageVariables
import net.bestia.zone.geometry.Vec3L

/**
 * Data transfer object to carry all needed data during a damage calculation.
 *
 * @author Thomas Felix
 */
sealed class BattleContext {
  abstract val usedAttack: BattleAttack
  abstract val attacker: BattleEntity
  abstract val damageVariables: DamageVariables
  abstract val weapon: Weapon

  abstract fun targetPosition(): Vec3L

  companion object {
    fun verifyGroundBattleContext(ctx: BattleContext): GroundBattleContext {
      return ctx as? GroundBattleContext
        ?: throw IllegalArgumentException("Context is not a GroundBattleContext")
    }
  }
}

/**
 * Context used when damage is calculated between entities.
 */
data class EntityBattleContext(
  override val usedAttack: BattleAttack,
  override val attacker: BattleEntity,
  override val weapon: Weapon,
  override val damageVariables: DamageVariables,
  val defender: BattleEntity,
) : BattleContext() {

  override fun targetPosition(): Vec3L {
    return defender.position
  }
}

/**
 * Context used if we choose to attack the ground.
 */
data class GroundBattleContext(
  override val usedAttack: BattleAttack,
  override val attacker: BattleEntity,
  override val weapon: Weapon,
  override val damageVariables: DamageVariables,
  val targetPosition: Vec3L
) : BattleContext() {
  override fun targetPosition(): Vec3L {
    return targetPosition
  }
}